File: src/Sekoia.Caching/Implementation/CacheEntry.cs
using System;
using System.Reactive;
using System.Reactive.Disposables;
using System.Reactive.Linq;
using System.Reactive.Subjects;
using System.Reactive.Threading.Tasks;
using System.Threading.Tasks;
using JetBrains.Annotations;

namespace Sekoia.Caching.Implementation
{
    internal class CacheEntry<T> : IDisposable
        where T : class
    {
        private static readonly Notification<CacheValue> _emptyNotification = Notification.CreateOnNext(new CacheValue(null, DateTimeOffset.MinValue));
        private static readonly Notification<CacheValue> _refreshNotification = Notification.CreateOnNext(new CacheValue(null, DateTimeOffset.MinValue));

        private readonly SerialDisposable _baseSubscription = new SerialDisposable();
        private readonly BehaviorSubject<Notification<CacheValue>> _latestValue = new BehaviorSubject<Notification<CacheValue>>(_emptyNotification);
        private readonly Func<TimeSpan> _maxAge;
        private readonly object _mutex = new object();
        private readonly Action _removeCacheEntry;

        private volatile Task<T> _currentElementRetrieverTask;

        private bool _refreshRequested;
        private Subject<Unit> _triggerFetchFromServer;

        public IObservable<T> Observable { get; }

        public CacheEntry([NotNull] Func<Task<T>> elementRetriever,
                          Func<TimeSpan> maxAge,
                          [NotNull] Action removeCacheEntry)
        {
            if (elementRetriever == null)
                throw new ArgumentNullException(nameof(elementRetriever));

            _maxAge = maxAge ?? throw new ArgumentNullException(nameof(maxAge));
            _removeCacheEntry = removeCacheEntry ?? throw new ArgumentNullException(nameof(removeCacheEntry));

            Observable = CreateObservable(elementRetriever);
        }

        public void Dispose()
        {
            _triggerFetchFromServer.Dispose();
            _baseSubscription.Dispose();
            _latestValue.Dispose();
        }

        public void Refresh()
        {
            if (_latestValue.HasObservers)
            {
                _latestValue.OnNext(_refreshNotification);
                _triggerFetchFromServer.OnNext(Unit.Default);
            }
            else
                _latestValue.OnNext(null);
        }

        public void SetCachedValue(T value)
        {
            _latestValue.OnNext(Notification.CreateOnNext(new CacheValue(value)));
        }

        /// <summary>
        /// Creates an Rx observable for the this instance. 
        /// </summary>
        /// <param name="elementRetriever">The retriever to use when fetching an element.</param>
        /// <returns>An observable that remains up to date when timing out and responds to external updates.</returns>
        private IObservable<T> CreateObservable(Func<Task<T>> elementRetriever)
        {
            _triggerFetchFromServer = new Subject<Unit>();

            // Creates an observable that retrieves an element and emits it, each time the
            // triggerFetchFromServer emits an event, regardless of value.
            var fetchFromServerObservable = _triggerFetchFromServer.Throttle(TimeSpan.FromMilliseconds(100))
                                                                   .Select(_ => SafelyCallElementRetrieverAsync(elementRetriever))
                                                                   .DistinctUntilChanged()
                                                                   .Select(t => t.ToObservable())
                                                                   .Switch()
                                                                   .Select(v => new CacheValue(v))
                                                                   .Materialize();

            _baseSubscription.Disposable = fetchFromServerObservable.Subscribe(v => _latestValue.OnNext(v));

            // Creates an observable only returns non-timed out valid values and errors.
            // Refreshes are triggered as necessary.
            var timeoutHandlingObservable = _latestValue.Where(v =>
                                                               {
                                                                   if (_refreshNotification.Equals(v))
                                                                       return false;
                                                                   if (v?.Kind == NotificationKind.OnCompleted)
                                                                       return false;
                                                                   if (v?.Kind == NotificationKind.OnError)
                                                                   {
                                                                       _removeCacheEntry();
                                                                       return true;
                                                                   }

                                                                   // If value hasn't timed out, use it
                                                                   if (!HasTimedOut(v?.Value.CacheTime ?? DateTimeOffset.MinValue))
                                                                       return true;

                                                                   // If value has timed out, ignore it and refresh if not currently refreshing
                                                                   if (_latestValue.HasObservers)
                                                                       _triggerFetchFromServer.OnNext(Unit.Default);

                                                                   return false;
                                                               })
                                                        .Dematerialize()
                                                        .Select(v => v.Value);

            return timeoutHandlingObservable;
        }

        private bool HasTimedOut(DateTimeOffset cacheTime)
        {
            return Age(cacheTime) > _maxAge();
        }

        private Task<T> SafelyCallElementRetrieverAsync(Func<Task<T>> elementRetriever)
        {
            lock (_mutex)
            {
                var currentElementRetrieverTask = _currentElementRetrieverTask;
                if (currentElementRetrieverTask != null)
                {
                    _refreshRequested = true;
                    return currentElementRetrieverTask;
                }


                var currentElementRetriverTask = elementRetriever();
                _currentElementRetrieverTask = currentElementRetriverTask;

                currentElementRetriverTask
                    .ContinueWith(t =>
                                  {
                                      lock (_mutex)
                                      {
                                          if (ReferenceEquals(currentElementRetriverTask, _currentElementRetrieverTask))
                                              _currentElementRetrieverTask = null;

                                          if (_refreshRequested)
                                          {
                                              _refreshRequested = false;
                                              Refresh();
                                          }
                                      }
                                  });

                return currentElementRetriverTask;
            }
        }

        private static TimeSpan Age(DateTimeOffset cacheTime)
        {
            return DateTimeOffset.Now - cacheTime;
        }

        private class CacheValue
        {
            public DateTimeOffset CacheTime { get; }
            public T Value { get; }

            public CacheValue(T value, DateTimeOffset? cacheTime = null)
            {
                Value = value;
                CacheTime = cacheTime ?? DateTimeOffset.Now;
            }
        }
    }
}

File: src/Sekoia.Caching.Widgets/sekoia-composition/Sekoia.Caching.Widgets/components/CacheBuilder/CacheEntry.ts
import rx = require("rx");
import helpers = require("sekoia/helpers");
import moment = require("moment");

class CacheEntry<TKey, TValue> implements rx.IDisposable
{
    private _isRefreshing: boolean;
    private _observable: Rx.Observable<TValue>;
    private _refreshFromElementRetrieverObserver: Rx.IObserver<any>;
    private _externalElementUpdateObserver: Rx.IObserver<ITimedValue<TValue>>;
    private _baseSubscription = new rx.SerialDisposable();

    public get id(): TKey
    {
        return this._id;
    }

    public get observable(): Rx.Observable<TValue>
    {
        return this._observable;
    }

    constructor(private _id: TKey,
                elementRetriever: () => Rx.Observable<TValue>,
                private _maxAgeInMilliseconds: () => number)
    {
        helpers.assert.parameterCannotBeNull(_id, "_id");
        helpers.assert.parameterCannotBeNull(elementRetriever, "elementRetriever");
        helpers.assert.parameterCannotBeNull(_maxAgeInMilliseconds, "_maxAgeInMilliseconds");

        // Used to indicate that an update from the retriever is necessary
        var refreshFromElementRetrieverSubject = new rx.ReplaySubject<any>(1);

        // Used to inject a new value from an external source (i.e. query cache)
        var externalElementUpdateObservable = new rx.BehaviorSubject<ITimedValue<TValue>>({
                value: null,
                cacheTime: moment({
                        year: 1,
                        month: 1,
                        day: 1
                    })
            });

        this._refreshFromElementRetrieverObserver = refreshFromElementRetrieverSubject;
        this._externalElementUpdateObserver = externalElementUpdateObservable;

        this._observable = this.createObservable(
            refreshFromElementRetrieverSubject,
            elementRetriever,
            externalElementUpdateObservable);
    }

    public refresh()
    {
        this._isRefreshing = true;
        this._externalElementUpdateObserver.onNext(null);
        this._refreshFromElementRetrieverObserver.onNext(1);
    }

    public dispose()
    {
        this._refreshFromElementRetrieverObserver.onCompleted();
        this._externalElementUpdateObserver.onCompleted();
        this._baseSubscription.dispose();
    }

    public setCachedValue(value: TValue)
    {
        this._isRefreshing = true;
        this._externalElementUpdateObserver
            .onNext({
                    value: value,
                    cacheTime: moment()
                });
    }

    /**
     * Creates an Rx observable for the this instance.
     * @param refreshFromElementRetrieverObservable An observable that emits an event each time the element should be fetched from the retriever.
     * @param elementRetriever The retriever to use when fetching an element.
     * @param externalElementUpdateObservable An observable that emits an event each time the element is updated externally.
     * @returns {} An observable that remains up to date when timing out and responds to external updates.
     */
    private createObservable(refreshFromElementRetrieverObservable: Rx.Observable<any>,
                             elementRetriever: () => Rx.Observable<TValue>,
                             externalElementUpdateObservable: Rx.BehaviorSubject<ITimedValue<TValue>>): Rx.Observable<TValue>
    {
        let externalSubscribers = 0;

        // Creates an observable that retrieves an element and emits it each time the
        // refreshFromElementRetrieverObservable emits an event, regardless of value.
        let retrieveElementOnRefreshObservable = refreshFromElementRetrieverObservable
            .selectSwitch(() =>
            {
                // If there aren't actually anyone using this entry we don't fetch it
                if (externalSubscribers === 0)
                {
                    // Reset refreshing to allow refreshing later
                    this._isRefreshing = false;
                    return rx.Observable.empty<ITimedValue<TValue>>();
                }

                return elementRetriever()
                    .select(v =>
                    {
                        return {
                                value: v,
                                cacheTime: moment()
                            };
                    });
            });

        // Creates an observable that emits an event each time a refresh is requested
        // or the element is updated externally.
        let retrievedOrExternallyUpdated = retrieveElementOnRefreshObservable
            .merge(externalElementUpdateObservable);

        const replayingObservable = retrievedOrExternallyUpdated
            .replay(undefined, 1); // Replay value if not first subscriber, starting with an empty value

        this._baseSubscription.setDisposable(replayingObservable.connect());

        // Creates an observable that simply wraps replayingObservable above
        // to perform a refresh when timed out, if not already refreshing.
        let timeoutHandlingObservable = replayingObservable
            .where(v =>
            {
                // If value hasn't timed out, use it
                if (v && !this.hasTimedOut(v.cacheTime))
                {
                    // Reset refreshing as a new value has been seen
                    this._isRefreshing = false;
                    return true;
                }

                // If value has timed out, ignore it and refresh if not currently refreshing
                if (!this._isRefreshing)
                    this.refresh();

                return false;
            })
            .select(v =>
            {
                return v.value;
            });

        const externalSubscriberTrackingObservable = rx.Observable
            .create<TValue>(o =>
            {
                // Increase refCount of external interesants
                externalSubscribers++;

                const subscription = timeoutHandlingObservable.subscribe(o);
                return rx.Disposable
                    .create(() =>
                    {
                        subscription.dispose();

                        // Increase refCount of external interesants
                        externalSubscribers--;
                        if (externalSubscribers === 0)
                            this._isRefreshing = false;
                    });
            });

        return externalSubscriberTrackingObservable;
    }

    private age(cacheTime: moment.Moment): number
    {
        return moment().diff(cacheTime);
    }

    private hasTimedOut(cacheTime: moment.Moment): boolean
    {
        return this.age(cacheTime) > this._maxAgeInMilliseconds();
    }
}

interface ITimedValue<TValue>
{
    value: TValue;
    cacheTime: moment.Moment;
}

export = CacheEntry;